zebra-consensus/src/block/subsidy/general.rs

```diff
@@ -0,0 +1,203 @@
+///
+/// [7.7]: https://zips.z.cash/protocol/protocol.pdf#subsidies
+fn slow_start_shift() -> Height {
+    Height(SLOW_START_INTERVAL.0 / 2)
+}
+
+/// The divisor used for halvings.
+///
+/// `1 << Halving(height)`, as described in [protocol specification §7.7][7.7]
+///
+/// [7.7]: https://zips.z.cash/protocol/protocol.pdf#subsidies
+pub fn halving_divisor(height: Height, network: Network) -> u64 {
+    if height < SLOW_START_INTERVAL {
+        panic!("can't verify before block {}", SLOW_START_INTERVAL.0)
+    }
+    let blossom_height = Blossom
+        .activation_height(network)
+        .expect("blossom activation height should be available");
+    if height >= blossom_height {
+        let scaled_pre_blossom_height =
+            (blossom_height - slow_start_shift()) as u64 * BLOSSOM_POW_TARGET_SPACING_RATIO;
+        let post_blossom_height = (height - blossom_height) as u64;
+        let halving_shift = (scaled_pre_blossom_height + post_blossom_height)
+            / (POST_BLOSSOM_HALVING_INTERVAL.0 as u64);
+        1 << halving_shift
+    } else {
+        let scaled_pre_blossom_height = (height - slow_start_shift()) as u64;
+        let halving_shift = scaled_pre_blossom_height / (PRE_BLOSSOM_HALVING_INTERVAL.0 as u64);
+        1 << halving_shift
+    }
+}
+
+/// `BlockSubsidy(height)` as described in [protocol specification §7.7][7.7]
+///
+/// [7.7]: https://zips.z.cash/protocol/protocol.pdf#subsidies
+pub fn block_subsidy(height: Height, network: Network) -> Result<Amount<NonNegative>, Error> {
+    if height < SLOW_START_INTERVAL {
+        panic!("can't verify before block {}", SLOW_START_INTERVAL.0)
+    }
+    let blossom_height = Blossom
+        .activation_height(network)
+        .expect("blossom activation height should be available");
+
+    let hd = halving_divisor(height, network);
+    if height >= blossom_height {
+        let scaled_max_block_subsidy = MAX_BLOCK_SUBSIDY / BLOSSOM_POW_TARGET_SPACING_RATIO;
+        // in future halvings, this calculation might not be exact
+        // in those cases, Amount division follows integer division, which truncates (rounds down) the result
+        Amount::try_from(scaled_max_block_subsidy / hd)
+    } else {
+        // this calculation is exact, because the halving divisor is 1 here
+        Amount::try_from(MAX_BLOCK_SUBSIDY / hd)
+    }
+}
+
+/// `MinerSubsidy(height)` as described in [protocol specification §7.7][7.7]
+///
+/// [7.7]: https://zips.z.cash/protocol/protocol.pdf#subsidies
+#[allow(dead_code)]
+pub fn miner_subsidy(height: Height, network: Network) -> Result<Amount<NonNegative>, Error> {
+    let canopy_height = Canopy
+        .activation_height(network)
+        .expect("canopy activation height should be available");
+    if height >= canopy_height {
+        panic!("Can't validate Canopy yet");
+    }
+    block_subsidy(height, network)? - founders_reward(height, network)?
+}
+
+#[cfg(test)]
+mod test {
```

Comment:
I made all these tests check testnet in f4ca8e16e3fbbff0a43fc2b19292be19c982cdb0 in oxarbitrage#29.